File: utilslib/src/main/java/com/ashlikun/utils/ui/keyboard/AutoHeightLayout.java
package com.ashlikun.utils.ui.keyboard;

import android.app.Activity;
import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Rect;
import android.util.AttributeSet;

import com.ashlikun.utils.ui.KeyboardUtils;

public class AutoHeightLayout extends SoftKeyboardSizeWatchLayout implements SoftKeyboardSizeWatchLayout.OnResizeListener {
    protected Context mContext;
    protected int mMaxParentHeight;
    protected int mSoftKeyboardHeight;
    protected boolean mConfigurationChangedFlag = false;

    public AutoHeightLayout(Context context) {
        this(context, null);
    }

    public AutoHeightLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public AutoHeightLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.mContext = context;
        mSoftKeyboardHeight = KeyboardUtils.getDefKeyboardHeight(mContext);
        addOnResizeListener(this);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        onSoftKeyboardHeightChanged(mSoftKeyboardHeight);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        if (mMaxParentHeight == 0) {
            mMaxParentHeight = h;
        }
    }

    public void updateMaxParentHeight(int maxParentHeight) {
        this.mMaxParentHeight = maxParentHeight;
        if (maxParentHeightChangeListener != null) {
            maxParentHeightChangeListener.onMaxParentHeightChange(maxParentHeight);
        }
    }

    @Override
    protected void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        mConfigurationChangedFlag = true;
        mScreenHeight = 0;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (mConfigurationChangedFlag) {
            mConfigurationChangedFlag = false;
            Rect r = new Rect();
            ((Activity) mContext).getWindow().getDecorView().getWindowVisibleDisplayFrame(r);
            if (mScreenHeight == 0) {
                mScreenHeight = r.bottom;
            }
            int mNowh = mScreenHeight - r.bottom;
            mMaxParentHeight = mNowh;
        }

        if (mMaxParentHeight != 0) {
            int heightMode = MeasureSpec.getMode(heightMeasureSpec);
            int expandSpec = MeasureSpec.makeMeasureSpec(mMaxParentHeight, heightMode);
            super.onMeasure(widthMeasureSpec, expandSpec);
            return;
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    public void onSoftPop(final int height) {
        if (mSoftKeyboardHeight != height) {
            mSoftKeyboardHeight = height;
            KeyboardUtils.setDefKeyboardHeight(mContext, mSoftKeyboardHeight);
            onSoftKeyboardHeightChanged(mSoftKeyboardHeight);
        }
    }

    @Override
    public void onSoftClose() {
    }

    /**
     * 键盘高度改变了
     */
    public void onSoftKeyboardHeightChanged(int height) {

    }

    private OnMaxParentHeightChangeListener maxParentHeightChangeListener;

    public interface OnMaxParentHeightChangeListener {
        void onMaxParentHeightChange(int height);
    }

    public void setOnMaxParentHeightChangeListener(OnMaxParentHeightChangeListener listener) {
        this.maxParentHeightChangeListener = listener;
    }
}

File: utilslib/src/main/java/com/ashlikun/utils/ui/keyboard/AutoHeightLayout.kt
package com.ashlikun.utils.ui.keyboard

import android.content.Context
import android.content.res.Configuration
import android.graphics.Rect
import android.util.AttributeSet
import com.ashlikun.utils.ui.extend.getActivity

typealias  OnMaxParentHeightChange = (height: Int) -> Unit

/**
 * @author　　: 李坤
 * 创建时间: 2021.12.14 11:34
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：自动适应软键盘高度的布局
 */
class AutoHeightLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : SoftKeyboardSizeWatchLayout(
    context, attrs, defStyleAttr
) {
    var maxParentHeightChange: OnMaxParentHeightChange? = null

    var maxParentHeight = 0
    var softKeyboardHeight = KeyboardUtils.getDefKeyboardHeight()
    var configurationChangedFlag = false


    init {
        onSoftPops.add {
            onSoftPop(it)
        }
        onSoftClose.add {
            onSoftClose()
        }
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        onSoftKeyboardHeightChanged(softKeyboardHeight)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        if (maxParentHeight == 0) {
            maxParentHeight = h
        }
    }

    fun updateMaxParentHeight(maxParentHeight: Int) {
        this.maxParentHeight = maxParentHeight
        maxParentHeightChange?.invoke(maxParentHeight)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        configurationChangedFlag = true
        screenHeight = 0
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (configurationChangedFlag) {
            configurationChangedFlag = false
            val r = Rect()
            context.getActivity()?.window?.decorView?.getWindowVisibleDisplayFrame(r)
            if (screenHeight == 0) {
                screenHeight = r.bottom
            }
            val mNowh = screenHeight - r.bottom
            maxParentHeight = mNowh
        }
        if (maxParentHeight != 0) {
            val heightMode = MeasureSpec.getMode(heightMeasureSpec)
            val expandSpec = MeasureSpec.makeMeasureSpec(maxParentHeight, heightMode)
            super.onMeasure(widthMeasureSpec, expandSpec)
            return
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    open fun onSoftPop(height: Int) {
        if (softKeyboardHeight != height) {
            softKeyboardHeight = height
            KeyboardUtils.setDefKeyboardHeight(context, softKeyboardHeight)
            onSoftKeyboardHeightChanged(softKeyboardHeight)
        }
    }

    open fun onSoftClose() {}

    /**
     * 键盘高度改变了
     */
    fun onSoftKeyboardHeightChanged(height: Int) {}


}